src/lem/constrainer.rs

```diff
@@ -236,98 +662,157 @@ impl LEM {
-                        } else {
-                            // If `branch_path_info` is None, we just do regular constraining
-                            alloc_ptr_computed.enforce_equal(
-                                &mut cs.namespace(|| format!("enforce equal for {output_name}")),
+                        alloc_ptr_computed
+                            .implies_ptr_equal(
+                                &mut cs
+                                    .namespace(|| format!("enforce imply equal for {output_name}")),
+                                &concrete_path,
                                 &alloc_ptr_expected,
-                            );
-                        }
+                            )
+                            .with_context(|| "couldn't constrain `implies_ptr_equal`")?;
                     }
                 }
                 _ => todo!(),
             }
         }
+
         if num_inputized_outputs != 3 {
             return Err(anyhow!("Couldn't inputize the right number of outputs"));
         }
+
+        if let Some(max_slots_indices) = max_slots_allowed {
+            if max_slots_indices.hash2_idx > hash_slots.hash2_data.max_slots {
+                bail!(
+                    "Too many slots allocated for Hash2/Unhash2: {}, {}",
+                    max_slots_indices.hash2_idx,
+                    hash_slots.hash2_data.max_slots
+                );
+            }
+            if max_slots_indices.hash3_idx > hash_slots.hash3_data.max_slots {
+                bail!(
+                    "Too many slots allocated for Hash3/Unhash3: {}, {}",
+                    max_slots_indices.hash2_idx,
+                    hash_slots.hash2_data.max_slots
+                );
+            }
+            if max_slots_indices.hash4_idx > hash_slots.hash4_data.max_slots {
+                bail!(
+                    "Too many slots allocated for Hash4/Unhash4: {}, {}",
+                    max_slots_indices.hash2_idx,
+                    hash_slots.hash2_data.max_slots
+                );
+            }
+        }
+
+        Self::create_slot_constraints(
+            cs,
+            &alloc_ptrs,
+            hash_slots,
+            store,
+            alloc_manager,
+        )?;
+
         Ok(())
     }
+
+    #[inline]
+    pub fn constrain<F: LurkField, CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        alloc_manager: &mut AllocationManager<F>,
+        store: &mut Store<F>,
+        valuation: &Valuation<F>,
+    ) -> Result<()> {
+        self.constrain_aux(cs, alloc_manager, store, valuation, None)
+    }
+
+    #[inline]
+    pub fn constrain_limited<F: LurkField, CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        alloc_manager: &mut AllocationManager<F>,
+        store: &mut Store<F>,
+        valuation: &Valuation<F>,
+        limits: &SlotsIndices,
+    ) -> Result<()> {
+        self.constrain_aux(cs, alloc_manager, store, valuation, Some(limits))
+    }
```

Comment:
Is it necessary to introduce two new aliases, or can we just :
- edit `constrain_aux` into a public `constrain_limited` which takes some `<T: Into<Option<&SlotIndices>>>` for its last argument (hence callable with `self.constrain_limited(cs, alloc_manager, store, valuation, limits)`), 
- add a public `constrain` that calls `self.constrain_limited(cs, alloc_manager, store, valuation, None)`,
- delete the rest?